docs/formatting.md

```diff
@@ -170,45 +225,101 @@ one (254814599) two (656421459) three (-1117028319)
 
-The following example changes the default for `Rectangle` to `text/plain`.
+```csharp
+[TypeFormatterSource(typeof(MyFormatterSource))]
+public class MyTypeWithCustomFormatting
+{
+}
+```
+
+A formatter source specified by the `TypeFormatterSourceAttribute` must implement `ITypeFormatterSource` and must have an empty constructor. It does not need to be a public type. Here's an example implementation:
 
 ```csharp
-using System.Drawing;
+internal class MyFormatterSource : ITypeFormatterSource
+{
+    public IEnumerable<ITypeFormatter> CreateTypeFormatters()
+    {
+        return new ITypeFormatter[]
+        {
+            new PlainTextFormatter<MyTypeWithCustomFormatting>(context => 
+              $"Hello from {nameof(MyFormatterSource)} using MIME type text/plain"),
+            new HtmlFormatter<MyTypeWithCustomFormatting>(context => 
+              $"Hello from {nameof(MyFormatterSource)} using MIME type text/html")
+        };
+    }
+}
+```
 
-Formatter.SetPreferredMimeTypesFor(typeof(Rectangle), "text/plain");
+As the example above shows, a formatter source can return multiple formatters. These can be formatters for different MIME types and also for different .NET types.
 
-new Rectangle
-{   
-    Height = 50, 
-    Width = 100
+A formatter source can also specify the preferred MIME types for your type:
+
+```csharp
+[TypeFormatterSource(
+  typeof(MyFormatterSource), 
+  PreferredMimeTypes = new[] { "text/html", "application/json" })]
+public class MyTypeWithCustomFormatting
+{
 }
 ```
 
-```console
-Rectangle
-  Location: Point
-    IsEmpty: True
-    X: 0
-    Y: 0
-  Size: Size
-    IsEmpty: False
-    Width: 100
-    Height: 50
-  X: 0
-  Y: 0
-  Width: 100
-  Height: 50
-  Left: 0
-  Top: 0
-  Right: 100
-  Bottom: 50
-  IsEmpty: False
+Specifying several MIME types results in a call to `Formatter.SetPreferredMimeTypesFor`, as described under [Preferred MIME types](#preferred-mime-types). The behavior is equivalent.
+
+If you don't want to take a dependency on `Microsoft.DotNet.Interactive.Formatting`, there's another to way to provide custom formatting for your types. This can be the simpler approach if you're building a library and you'd like to provide an augmented experience in notebooks but you don't need the other tools available to .NET Interactive extensions, such as magic commands and custom kernels.
+
+With this approach, you declare your own `TypeFormatterSourceAttribute` and apply it. The shape of the class should match the `TypeFormatterSourceAttribute` defined in `Microsoft.DotNet.Interactive.Formatting`. The `TypeFormatterSourceAttribute` class that you define will be recognized by name. Like the formatter source, it can be declared as internal so that it has no impact on your public API surface.
+
+Here's the complete definition, which you can copy into your project:
+
+```csharp
+[AttributeUsage(AttributeTargets.Class)]
+internal class TypeFormatterSourceAttribute : Attribute
```

Comment:
Would it be better to add a link to the source file for `TypeFormatterSourceAttribute` here instead (so that it always reflects the latest API)?

Reply:
It won't change, and the readability is clearer to have it on the doc.